feat(CategoryTheory): The Preliminaries for Locally Cartesian Closed Categories

sinhp opened the pull request:
See https://github.com/leanprover-community/mathlib4/pull/21525

This PR defines the basic preliminaries for defining locally cartesian closed categories (LCCCs). 

First, we develop a computable implementation of pullbacks, by introducing a new type-class `ChosenPullback`. The non-computable `HasPullbacksAlong` and `HasPullbacks` predicates yield instances of `ChosenPullback` using global choice, but interestingly in the category of types every morphism has chosen pullbacks. Also, cartesian monoidal categories, morphisms to the terminal object and the product projections have chosen pullbacks. We prove that `ChosenPullback` has good closure properties, e.g., isos have chosen pullbacks, and composition of morphisms with chosen pullbacks have chosen pullbacks.
 
Separately, we prove that pull-push along a morphism with chosen pullbacks gives the (cartesian) binary product of that morphism, seen as an object of a slice category, and any other object of the same slice category. In fact, we also prove the stronger statement, namely that the pull-push composition `(Over.ChosenPullback.pullback Z.hom) ⋙ (Over.map Z.hom)` is naturally isomorphic to the right tensor product functor `_ × Y` in `Over X`. _This is going to be crucial in our mate-based approach to LCCCs_. 

Also, using the calculus of mates we define certain natural isomorphisms involving `Over.star` and `Over.pullback` which will be used in defining the right adjoint to the pullback functor in the development of LCCCs.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

sinhp committed:
the preliminaries for locally cartesian closed categories (Sigma, Reindex, pull-push)

(Sigma=push, Reindex=pull, pull-push = cartesian product)

sinhp committed:
Merge remote-tracking branch 'refs/remotes/sina/lccc-prelim' into lccc-prelim

sinhp committed:
prelims for lcccs

sinhp committed:
fix for HasPullbacksAlong

sinhp committed:
parentheses and simp

sinhp committed:
Update Mathlib.lean

sinhp committed:
not needed assumption

sinhp committed:
rm simp

sinhp committed:
correcting attribution

sinhp committed:
Update Prelim.lean

sinhp committed:
no data using tactics

sinhp committed:
not using Equivalence.mk, instead construct equiv directly

sinhp committed:
IsTerminal instead of HasTerminal

sinhp committed:
lower case sigma and reindex

sinhp committed:
rm equivalenceRightAdjointIsoInverse

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Prelim.lean:
```diff
@@ -0,0 +1,379 @@
+  map {X Y} f := Over.homMk f
+
+/-- The slice category over the terminal object is equivalent to the original category. -/
+def equivOverTerminal (X : C) (h : IsTerminal X) : Over (X) ≌ C where
```
Could you move this to the `IsTerminal` namespace? Also, the `inverse` functor could probably be "inlined" in the definition of the equivalence. Could you also add the `simps` attribute?

sinhp commented:
@joelriou 
Another thing to note that is that `a : Over.ChosenPullback f` and `b : Over.ChosenPullback g` gives `c : Over.ChosenPullback (f.comp g)` when the latter makes sense. However the pullback map `c.pullback _` "chooses" is not the equal to `a.pullback (b.pullback _)` but only isomorphic. so we the term is "Chosen" needs to be understood in "up to iso" sense.

joelriou commented:
> In particular if you have `f` an inverse for `g`, then `f.comp g = id` will have two induced instances `Over.ChosenPullback id`, one from the identity and the other from composition, and there is no way to show they are the same. Won't this be problematic?
`Over.ChosenPullback.id/comp` should only be definitions, not instances...

sinhp commented:
Sounds good, I'll try to implement these in the next commit. Hopefully the file will become much shorter.

sinhp committed:
ChosenPullback

sinhp committed:
ChosenPullback (wip)

sinhp commented:
- removed `sigma` and `reindex` defs and their API 
- introduced notation instead using macros, now APIs is minimal 
- promoted some constructions to `ChosenPullback`. Proved that maps to terminal and the basic product projections have always chosen pullbacks  in any cartesian monoidal category.

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Prelim.lean:
```diff
@@ -0,0 +1,452 @@
+  mapPullbackAdj := Over.mapPullbackAdj f
+
+/-- The identity morphism has a chosen pullback. -/
+def id {X : C} : ChosenPullback (𝟙 X) where
```
```suggestion
@[simps]
def id {X : C} : ChosenPullback (𝟙 X) where
```

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Prelim.lean:
```diff
@@ -0,0 +1,452 @@
+/-- The identity morphism has a chosen pullback. -/
+def id {X : C} : ChosenPullback (𝟙 X) where
+  pullback := 𝟭 _
+  mapPullbackAdj := Adjunction.ofNatIsoLeft  (Adjunction.id) (Over.mapId _).symm
```
```suggestion
  mapPullbackAdj := (Adjunction.id).ofNatIsoLeft (Over.mapId _).symm
```

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Prelim.lean:
```diff
@@ -0,0 +1,452 @@
+    [i₁ : ChosenPullback f] [i₂ : ChosenPullback g] : ChosenPullback (g ≫ f) where
+  pullback := i₁.pullback ⋙ i₂.pullback
+  mapPullbackAdj := Adjunction.ofNatIsoLeft  (i₂.mapPullbackAdj.comp i₁.mapPullbackAdj)
+    (Over.mapComp g f).symm
```
```suggestion
  mapPullbackAdj := (i₂.mapPullbackAdj.comp i₁.mapPullbackAdj).ofNatIsoLeft
    (Over.mapComp g f).symm
```

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Prelim.lean:
````diff
@@ -0,0 +1,452 @@
+```
+-/
+@[simp]
+theorem isPullback (Y Z : Over X) [ChosenPullback Y.hom] :
````
I would rather see an unbundled API, which mimics both `CartesianMonoidalCategory` and `pullback.*`:
```lean
section

variable {Y Z X : C} (f : Y ⟶ X) (g : Z ⟶ X) [ChosenPullback g]

abbrev pullbackObj := ((pullback g).obj (Over.mk f)).left

abbrev fst : pullbackObj f g ⟶ Y := ((mapPullbackAdj g).counit.app (Over.mk f)).left

abbrev snd : pullbackObj f g ⟶ Z := ((pullback g).obj (Over.mk f)).hom

@[reassoc]
lemma condition : fst f g ≫ f = snd f g ≫ g :=
  Over.w ((mapPullbackAdj g).counit.app (Over.mk f))

variable {f g} in
@[ext]
lemma hom_ext {W : C} {φ₁ φ₂ : W ⟶ pullbackObj f g} (h₁ : φ₁ ≫ fst _ _ = φ₂ ≫ fst _ _)
    (h₂ : φ₂ ≫ snd _ _ = φ₂ ≫ snd _ _) :
    φ₁ = φ₂ := by
  sorry

section

variable {f g} {W : C} (a : W ⟶ Y) (b : W ⟶ Z) (h : a ≫ f = b ≫ g)

def lift :
    W ⟶ pullbackObj f g :=
  (((mapPullbackAdj g).homEquiv (Over.mk b) (Over.mk f)) (Over.homMk a)).left

@[reassoc (attr := simp)]
lemma lift_fst : lift a b h ≫ fst f g = a := by
  sorry

@[reassoc (attr := simp)]
lemma lift_snd : lift a b h ≫ snd f g = b := by
  simp [lift]

theorem isPullback {Y Z X : C} (f : Y ⟶ X) (g : Z ⟶ X) [ChosenPullback g] :
    IsPullback (fst f g) (snd f g) f g where
  w := condition f g
  isLimit' :=
    ⟨PullbackCone.IsLimit.mk _ (fun s ↦ lift s.fst s.snd s.condition)
      (by simp) (by simp) (by aesop)⟩

end

end
```

sinhp committed:
unbundle

sinhp committed:
updating some doc strings

sinhp committed:
naming

sinhp committed:
fix linter errors

sinhp committed:
defeq check for lift

sinhp committed:
Merge branch 'master' into lccc-prelim

FlAmmmmING committed:
feat(Combinatorics/Enumerative/Bell.lean): define standard Bell number (#26790)



Co-authored-by: jasin <219660779@qq.com>

joelriou commented:
Could you make a PR with only:
* the basic `ChosenPullback` API, in a file `CategoryTheory.LocallyCartesianClosed.ChosenPullback`;
* the definition of this instance on `Type` in a separate file `CategoryTheory.LocallyCartesianClosed.Type`.

urkud committed:
feat(Topology): infinite T2 spaces (#30661)

joelriou committed:
chore(CategoryTheory/EssentiallySmall): adding an instance (#30534)

Before this PR, `example (C : Type w) [SmallCategory C] : EssentiallySmall.{w} C := inferInstance` would not work. We fix this by making `essentiallySmall_of_small_of_locallySmall` an instance.

Consequently, we need to make certain universes explicit in a few proofs. Some instances in `Sites.Equivalence` are transformed into lemmas because it seems they shadowed more basic instances.

smmercuri committed:
feat(UniformSpace/Completion): add `UniformSpace.Completion.mapEquiv` (#30574)

If `α ≃ᵤ β` then `Completion α ≃ᵤ Completion β`.

loefflerd committed:
feat(Topology/Algebra/Order/Archimedean): subgroups cyclic iff discrete (#30471)

An add subgroup of the reals (or similar groups) is discrete iff it's cyclic, and related statements.

vasnesterov committed:
feat(Analysis/Calculus): Taylor series converges to function on whole ball (#26267)

* Prove `FormalMultilinearSeries.AnalyticOnNhd`: the sum of series is analytic on its ball of convergence.
* Prove `AnalyticOn.hasFPowerSeriesOnBall`: a variant of `AnalyticAt.hasFPowerSeriesAt` which gives convergence to the function on the ball of convergence.

alreadydone committed:
feat(Algebra): `SemimoduleCat`, category of modules over semiring (#31023)

+ Introduce SemimoduleCat, the category of semimodules (mathlib's Module) over a Semiring. The new file ModuleCat/Semi.lean is copied from ModuleCat/Basic.lean. The additive/linear instances would require #28826.

emilyriehl committed:
feat(AlgebraicTopology): the homotopy category functor preserves products (#25780)

This proves that the homotopy category construction defines a monoidal functor from simplicial sets to categories. In particular, simplicially enriched categories have quotient categorically enriched categories, which might be thought of as "homotopy bicategories."

Co-authored-by: Dominic Verity <dom.verity@gmail.com>, Bhavik Mehta <bhavikmehta8@gmail.com>, Thomas Murrills <thomasmurrills@gmail.com>, and Andrew Yang <the.erd.one@gmail.com>.
Co-authored-by: Mario Carneiro <di.gama@gmail.com>
Co-authored-by: emilyriehl <eriehl@jhu.edu>

joelriou committed:
feat(Algebra/Homology): Ext modules (#27416)

In this PR, we show that if `C` is a `R`-linear abelian category, then there is a `R`-module structure on the groups `Ext X Y n` for `X` and `Y` in `C` and `n : ℕ`.

joelriou committed:
feat(AlgebraicTopology/SimplicialSet): pairings of non degenerate simplices, following Sean Moss (#28332)

In this PR, we introduce the notion of pairing for a subcomplex of a simplicial set. Following the work of Sean Moss,  *Another approach to the Kan-Quillen model structure*, this is an essential tool in the study of (inner) anodyne extensions.

joelriou committed:
feat(CategoryTheory): strong generators (#29519)

In this PR, we introduce a condition `IsStrongGenerator P` on `P : ObjectProperty C` which is stronger than `IsSeparating P` as we also require that for any proper subobject `A ⊂ X`, there exists a morphism `G ⟶ X` from an
object satisfying `P` which does not factor through `A`.

In case coproducts exists, we give a characterization of `IsStrongSeparator P` in terms of the existence, for each object `X : C` of an extremal epi to `X` from a coproduct of objects in `S`. 

In future works (#30241), this shall be used in order to characterize locally `κ`-presentable categories as cocomplete categories that have a strong generator consisting of `κ`-presentable objects.

grunweg committed:
feat: add mfderiv_prod and mfderiv_prodMap (#29589)

Part of my bordism theory project: these will be needed for https://github.com/leanprover-community/mathlib4/pull/22059.

Updated version of #22642.

joelriou committed:
feat(CategoryTheory/ObjectProperty): closure of a property under certain limits (#29854)

In this PR, given a property `P` of objects in a category `C` and family of categories `J : α → Type _`, we introduce the closure `P.limitsClosure J` of `P` under limits of shapes `J a` for all `a : α`, and under certain smallness circumstances, we show that its essentially small.

- [x] depends on: #29851
- [x] depends on: #29881
- [x] depends on: #29903
- [x] depends on: #29849
- [x] depends on: #29843

joelriou committed:
feat(CategoryTheory): Guitart exact squares and Kan extensions (#30382)

Given a Guitart exact square `w : T ⋙ R ⟶ L ⋙ B`, we show that pointwise left Kan extensions along `R` give, by composition, pointwise left Kan extensions along `L`.

joelriou committed:
feat(CategoryTheory): κ-filtered categories are stable under products (#30634)

Binary products and general products of `κ`-filtered categories are `κ`-filtered.

emilyriehl committed:
Feat(Logic): basic properties of boolean negation (#31028)

We show that `not : Bool -> Bool` is an involutive equivalence and derive some immediate consequences.

Co-authored-by: Kenny Lau <kc_kennylau@yahoo.com.hk>
Co-authored-by: Wrenna Robson <wren.robson@gmail.com>

YaelDillies committed:
feat: relation-separated sets (#23920)

Define a notion of separation of a set relative to a relation. This will be used to unify metric and dynamical separation.

From MiscYD and LeanAPAP

joelriou committed:
feat(Topology/Sheaves): the equivalence `Over U ≌ Opens U` for `U : Opens X` (#30415)

joelriou committed:
feat(SetTheory): more API for HasCardinalLT (#30633)

We add stability lemmas for `HasCardinalLT X κ` for sigma types, binary products, etc when `κ` is regular (or at least infinite).

joelriou committed:
feat(CategoryTheory/Limits): precomposition with final functors and ColimitType (#30779)

We obtain a dual result to #30403 for the interaction between final functors and colimits of functors to types. This is phrased using the universe generic `Functor.ColimitType`. (We also improve the proof of #30403)

bryangingechen committed:
ci: fix concurrency group (#31032)

Multiple build runs from the same PR don't seem to be getting canceled. There's a suspicious line break in the `concurrency` option which might be confusing GitHub and causing this.

cf. [#mathlib4 > github action bandwidth @ 💬](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/github.20action.20bandwidth/near/547550820).

alreadydone committed:
feat(Algebra/Category): symmetric monoidal structure on `SemimoduleCat` (#30638)

+ Construct the symmetric monoidal structure on SemimoduleCat, and transport it to the same structure on ModuleCat (in a way preserving pre-existing defeqs) using the equivalence `ModuleCat.equivalenceSemimoduleCat` between the categories. Just a few proofs in the Representation library need to be fixed (in a straightforward way). Public API for ModuleCat is duplicated to SemimoduleCat.

harahu committed:
doc(Algebra): a collection of nontrivial typo fixes (#30749)

This is a collection of typo fixes suggested to me by Codex. I decided to split these changes out of #30669, because I feel none of these are completely trivial to review, mostly because they require inspecting the context of the text as opposed to just the text itself.

I've manually reviewed all of them to the best of my abilities and I'm reasonably sure they are all correct, and that they represent an improvement over the status quo. However, I am not an expert in Algebra, so please review with care.

I suspect that there might still be quite a wide span in the difficulty of reviewing the changes in this PR, but I don't have a strong intuition for which changes herein might still be seen as relatively easy to review. I'm open to splitting this PR further should any reviewer think that is a good idea.

Ruben-VandeVelde committed:
chore: remove some unused variables (#31017)

joelriou committed:
feat(CategoryTheory): prestacks (#30177)

Let `C` be a category and `F : Pseudofunctor (LocallyDiscrete Cᵒᵖ) Cat`. Given `S : C`, and objects `M` and `N` in `F.obj (.mk (op S))`, we define a presheaf of types `F.presheafHom M N` on the category `Over S`: its sections on an object `T : Over S` corresponding to a morphism `p : X ⟶ S` are the type of morphisms `p^* M ⟶ p^* N`. We shall say that `F` satisfies the descent of morphisms for a Grothendieck topology `J` (i.e. `F` is a prestack) if these presheaves are all sheaves.

Co-authored-by: Christian Merten [christian@merten.dev](mailto:christian@merten.dev)

xroblot committed:
chore(NumberField/Cyclotomic): add statements for the case of power one (#30996)

luigi-massacci committed:
feat: add API about the support of functions in ContDiffMapSupportedIn (#30199)

Add simple API lemmas about the support of functions in `ContDiffMapSupportedIn`.

Co-authored-by: Michael Rothgang <rothgang@math.uni-bonn.de>
Co-authored-by: Anatole Dedecker <anatolededecker@gmail.com>

RemyDegenne committed:
feat: convergence in distribution and continuous mapping theorem (#30540)

- Define convergence in distribution of random variables: this is the weak convergence of their laws.
- Prove the continuous mapping theorem for convergence in distribution and continuous functions.

Co-authored-by: Remy Degenne <remydegenne@gmail.com>

kebekus committed:
feat: elementary lemmas aiding computations with `log⁺`, Mark II (#30729)

Add several elementary lemmas aiding computations with log⁺. Improve the docstrings a little.

joelriou committed:
feat(CategoryTheory/Final): improve assumptions for `final_fst` (#30655)

We show that the projection `C × D ⥤ C` is a final (or initial) functor when `D` is connected. (Before this PR, it was proven under the stronger assumption that `D` was (co)filtered.)

joelriou committed:
feat(Order/Category): partial orders with order embeddings as morphisms (#30693)

This category will be relevant in the study of accessible categories. I am making Johan Commelin a co-author as the new file `PartOrdEmb.lean` is essentially a copy-paste of `PartOrd.Lean`.

Co-authored-by: Johan Commelin <johan@commelin.net>

kebekus committed:
doc: fix trivial typo in docstring (#31045)

Fixes a typo in a docstring.

vasnesterov committed:
feat(Analysis/Complex): `enorm` versions for `OperatorNorm.lean` (#31049)

Add `enorm` versions of theorems in `OperatorNorm.lean` as it currently has only `norm` and `nnnorm` versions.

grunweg committed:
feat: tag `IsInducing` and `Is{Open,Closed,}Embedding` with `fun_prop` (#29082)

I have not tested if this is usable in practice; perhaps it is not.

joelriou committed:
feat(CategoryTheory/Adjunction): more mate compatibilities (#30335)

We obtain various lemmas about adjunctions between functors by computing `conjugateEquiv` in case of associators, unitors, and whiskerings. We deduce a way to obtain compatibilites with respect to composition of functors of left adjoints assuming that we have these compatibilites for right adjoints. In #30318, this shall be applied to the study of the compatibilities with respect to composition of the pullback functors for presheaves of modules.

grunweg committed:
chore: make MDifferentiableAt.mfderiv_prod an alias. (#31048)

It duplicates a lemma above, but having this name is nice for discoverability.

urkud committed:
feat(LocallyFinite): add `of_comp_surjective` (#30662)

smmercuri committed:
feat(InfinitePlace/Ramification): `LiesOver` class for absolute values and derived results on infinite places (#27978)

If `w` is an absolute value on `L/K` and `v` is an absolute value of `K` then `w.LiesOver v` is the class defining the property that `v` is the restriction of `w` to `K`.

This PR continues the work from #24881.

Original PR: https://github.com/leanprover-community/mathlib4/pull/24881

mcdoll committed:
feat(docs/undergrad): add calculus and Schwartz function results (#31052)

gasparattila committed:
feat(Topology/Sets): `Singleton` instances for `(Nonempty)Compacts` (#31035)

mcdoll committed:
chore(Analysis/Distribution): split off functions and measures of temperate growth  (#30553)

The Schwartz function file reached 1500 lines of code, so it really needs splitting.

The only other change is that I added namespaces and moved one lemma into the `MeasureTheory.Measure` namespace to allow for dot-notation.

kckennylau committed:
chore: deprecate frobenius_xxx lemmas (#31014)

e.g. `frobenius_add` should be replaced with `map_add _`.

grunweg committed:
chore: mark measure_union_ne_top and friends with finiteness (#30967)

In some cases, add `≠ ∞` versions of existing `< ∞` lemmas, and tag those with finiteness.
Use this to golf one proof `by measurability` using finiteness, which is the morally correct proof.
Motivated by #30966.

sinhp committed:
Merge branch 'lccc-prelim' of https://github.com/sinhp/mathlib4 into lccc-prelim
